plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,457 @@
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"log"
+	"strings"
+)
+
+const (
+	// plugin is used to identify ourselves in log output
+	plugin = "openstack"
+)
+
+// serviceType is an OpenStack service type
+type serviceType string
+
+const (
+	volumeV2Service serviceType = "volumev2"
+)
+
+// tagMap maps a tag name to value.
+type tagMap map[string]string
+
+// fieldMap maps a field to an arbitrary data type.
+type fieldMap map[string]interface{}
+
+// serviceMap maps service id to a Service struct.
+type serviceMap map[string]services.Service
+
+// ContainsService indicates whether a particular service is enabled
+func (s serviceMap) ContainsService(t serviceType) bool {
+	for _, service := range s {
+		if service.Type == string(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// projectMap maps a project id to a Project struct.
+type projectMap map[string]projects.Project
+
+// hypervisorMap maps a hypervisor id a Hypervisor struct.
+type hypervisorMap map[int]hypervisors.Hypervisor
+
+// serverMap maps a server id to a Server struct.
+type serverMap map[string]servers.Server
+
+// flavorMap maps a flavor id to a Flavor struct.
+type flavorMap map[string]flavors.Flavor
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// volumeMap maps a volume id to a volume struct.
+type volumeMap map[string]volume
+
+// storagePoolMap maps a storage pool name to a StoragePool struct.
+type storagePoolMap map[string]schedulerstats.StoragePool
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint string
+	Domain           string
+	Project          string
+	Username         string
+	Password         string
+
+	// Locally cached clients
+	identity *gophercloud.ServiceClient
+	compute  *gophercloud.ServiceClient
+	volume   *gophercloud.ServiceClient
+
+	// Locally cached resources
+	services     serviceMap
+	projects     projectMap
+	hypervisors  hypervisorMap
+	flavors      flavorMap
+	servers      serverMap
+	volumes      volumeMap
+	storagePools storagePoolMap
+}
+
+// Description returns a description string of the input plugin and implements
+// the Input interface.
+func (o *OpenStack) Description() string {
+	return "Collects performance metrics from OpenStack services"
+}
+
+// sampleConfig is a sample configuration file entry.
+var sampleConfig = `
+  ## This is the recommended interval to poll.
+  interval = '60m'
+
+  ## The identity endpoint to authenticate against and get the
+  ## service catalog from
+  identity_endpoint = "https://my.openstack.cloud:5000"
+
+  ## The domain to authenticate against when using a V3
+  ## identity endpoint.  Defaults to 'default'
+  domain = "default"
+
+  ## The project to authenticate as
+  project = "admin"
+
+  ## The user to authenticate as, must have admin rights
```

Comment:
This seems unfortunate, can we do better than this (less privilege?)  Do all of the objects require admin?

Reply:
Allow me to jump in here - I've been using this plugin for a while - and have some additions queued to propose when this finally lands...
project = "admin" is typically the default project/tenant in a fresh Openstack install, so it makes a sensible default as all other values are user-defined.
user = "admin" - _does_ have to be an admin-level user, per the comment, because certain of the Openstack API's being invoked here are admin-only. I think we can assume that anyone who has deployed a large Openstack Cloud is aware of the privilege model.

Reply:
these are sensible defaults. Since any metrics component for openstack need to be able to read through across projects, admin privilege makes sense. We can provide an account that only has read only privilege. But account like that need to be created and configured with various policies across different components, with openstack's lax implementation of RBAC it could be difficult to achieve the ideal read only account.

Reply:
As pointed out, the APIs being polled require administrative privileges.

Reply:
Ok with me, you can mark this as resolved.